helix-core/src/main/java/org/apache/helix/messaging/handling/HelixStateTransitionHandler.java

```diff
@@ -233,7 +195,9 @@ void postHandleMessage() {
         _stateModelFactory.removeStateModel(resource, partitionKey);
       } else {
         // if the partition is not to be dropped, update _stateModel to the TO_STATE
-        _stateModel.updateState(toState);
```

Comment:
Question: Why removing the requested state needs this check? Don't we already need such a check?

Reply:
Originally `_stateModel` is only updated here at `postHandleMessage`. Now `TaskRunner` also updates `TaskRunner` on a different thread, so we need to check if `_stateModel` is already updated here before writing the new thread.  
Considering the following case. When controller sends a INIT->RUNNING for a certain task. `TaskRunner` is create in a separate thread and finishes executing the task super fast and update `_stateModel` to COMPLETED.  Now `postHandleMessage()` wants to change `_stateModel` to RUNNING, we should reject such change. If current `_stateModel` is different from message's from state, it means `_stateModel` is updated by `TaskRunner`.

Reply:
+1. Then what is the behavior for the rest of the process? Let's say we dont update it. Do we need to throw an exception for this message processing or we just drop the message and treated it as nothing happened? This could be a confusing behavior.

Reply:
I think if this is the case we  just don't update `_stateModel` or `CurrentState`. No exception will be thrown.

Reply:
@xyuanlu It will be confusing for the logic:
 
You have a message and finished the process ready to update current state. But you did not update the current state. Then you ignore it without anything.

My questions: 1) How you define this message status, success? or failed?
                         2) With this message processing, the application status could be changed internally. The real current state does not reflect the application real current state. Shall we still let it to be the state?

Reply:
1. The message status will still be success. Task did finished the state transition indicated by the message. For example, when participant receives a message INIT->RUNNING, task did started to run and completed. The code high lighted here only skips ZK update. The message handling part is not skipped.
2. We skip ZK update here to make sure the current state on ZK is the same as real task's current state. When task finished running and ZK current state is COMPLETED (Updated in TaskRunner). It does not make sense to change it to RUNNING (When postMessageHandler try to update ZK current state to the message's to state here).